content/blog/2016-07-23-pull-does-not-scale-or-does-it.md

```diff
@@ -0,0 +1,178 @@
+ingest data (especially persisting time series data on disk). To put some
+numbers behind this: a single beefy Prometheus server can easily store millions
+of time series, with a record of 800,000 incoming samples per second (as
+measured with real production metrics data at SoundCloud). This allows you to
+monitor over 10,000 machines from a single Prometheus server. The scaling
+bottleneck here has never been related to pulling metrics, but usually to the
+speed at which the Prometheus server can ingest the data into memory and then
+sustainably persist and expire data on disk/SSD.
+
+Also, although networks are pretty reliable these days, using a
+TCP-based pull approach makes doubly sure that metrics data arrives reliably,
+or that the monitoring system at least knows immediately when the metrics
+transfer fails due to a broken network.
+
+## Prometheus is not an event-based system
+
+Some monitoring systems are event-based. That is, they report each individual
+event (an HTTP request, an exception, you name it) to a central monitoring
+system immediately as it happens. This central system then either aggregates
+the events into metrics (StatsD is one example of this) or stores events
+individually for later processing (the ELK stack is an example of that). In
+such a system, pulling would be problematic indeed: the instrumented service
+would have to buffer events between pulls, and the pulls would have to happen
+incredibly frequently in order to simulate the same “liveness” of the
+push-based approach and not overwhelm event buffers.
+
+However, again, Prometheus is not an event-based monitoring system. You do not
+send raw events to Prometheus, nor can it store them. Prometheus is in the
+business of collecting aggregated time series data. That means that it's only
+interested in regularly collecting the current *state* of a given set of
+metrics, not the underlying events that led to the generation of those metrics.
+For example, an instrumented service would not send a message about each HTTP
+request to Prometheus as it is handled, but would simply count up those
+requests in memory.  This can happen hundreds of thousands of times per second
+without causing any monitoring traffic. Prometheus then simply asks the service
+instance every 15 or 30 seconds (or whatever you configure) about the current
+counter value and stores that value together with the scrape timestamp as a
+sample. Other metric types, such as gauges, histograms, and summaries, are
+handled similarly. The resulting monitoring traffic is low, and the pull-based
+approach also does not create problems in this case.
+
+## But now my monitoring needs to know about my service instances!
+
+With a pull-based approach, your monitoring system needs to know which service
+instances exist and how to connect to them. Often people are worried about the
+extra configuration this requires on the side of the monitoring system.
```

Comment:
"on the monitoring system side"

it'd be "on the part of the monitoring system"